File: examples/llama.android.bench/app/src/main/java/com/example/llama/bench/DeviceFarmBenchConfig.kt
package com.example.llama.bench

import android.content.Context
import org.json.JSONObject

data class DeviceFarmBenchConfig(
    val automation: Boolean = false,
    val modelPath: String? = null,
    val modelDownloadUrl: String? = null,
    val nGpuLayers: Int? = null,
    val repetitions: Int? = null,
    val skipDownload: Boolean? = null,
) {
    companion object {
        fun loadFromAssets(context: Context): DeviceFarmBenchConfig? {
            return try {
                context.assets.open("devicefarm_bench.json").use { stream ->
                    val json = JSONObject(stream.bufferedReader().readText())
                    DeviceFarmBenchConfig(
                        automation = json.optBoolean("automation", false),
                        modelPath = json.optString("model_path", null),
                        modelDownloadUrl = json.optString("model_download_url", null).takeIf { it.isNotEmpty() },
                        nGpuLayers = if (json.has("n_gpu_layers")) json.getInt("n_gpu_layers") else null,
                        repetitions = if (json.has("repetitions")) json.getInt("repetitions") else null,
                        skipDownload = if (json.has("skip_download")) json.getBoolean("skip_download") else null,
                    )
                }
            } catch (_: Exception) {
                null
            }
        }
    }
}

File: examples/llama.swiftui_bench/llama.swiftui_bench/Models/DeviceFarmBenchConfig.swift
import Foundation

struct DeviceFarmBenchConfig: Decodable {
    let automation: Bool?
    let modelPath: String?
    let modelDownloadURL: String?
    let nGpuLayers: Int?
    let repetitions: Int?
    let skipDownload: Bool?

    enum CodingKeys: String, CodingKey {
        case automation
        case modelPath = "model_path"
        case modelDownloadURL = "model_download_url"
        case nGpuLayers = "n_gpu_layers"
        case repetitions
        case skipDownload = "skip_download"
    }

    static func loadFromBundle() -> DeviceFarmBenchConfig? {
        guard let url = Bundle.main.url(forResource: "devicefarm_bench", withExtension: "json"),
              let data = try? Data(contentsOf: url) else {
            return nil
        }
        return try? JSONDecoder().decode(DeviceFarmBenchConfig.self, from: data)
    }
}
